irohad/torii/impl/command_service_transport_grpc.cpp

```diff
@@ -234,21 +241,32 @@ namespace torii {
           last_tx_status = status;
           return result;
         })
-        .subscribe(subscription,
-                   [this, &response_writer, &client_id](const auto &response) {
-                     if (response_writer->Write(response)) {
-                       log_->debug("status written, {}", client_id);
-                     }
-                   },
-                   [&](std::exception_ptr ep) {
-                     log_->error("something bad happened, client_id {}",
-                                 client_id);
-                   },
-                   [&] { log_->debug("stream done, {}", client_id); });
+        .subscribe(
+            subscription,
+            [this, &subscription, &response_writer, &client_id, &stream_failed](
+                const auto &response) {
+              // TODO [IR-249] akvinikym 23.01.19: remove the mutex after
+              // ensuring only one thread can be here
+              std::lock_guard<std::mutex> lg{stream_write_mutex_};
+              if (not stream_failed and response_writer->Write(response)) {
+                log_->debug("status written, {}", client_id);
+              } else {
+                log_->error("write to stream has failed to client {}",
+                            client_id);
+                stream_failed = true;
+                subscription.unsubscribe();
+              }
+            },
+            [&](std::exception_ptr ep) {
+              log_->error("something bad happened, client_id {}", client_id);
+              stream_failed = true;
+              subscription.unsubscribe();
```

Comment:
This is not required, since the observable will not emit any events after an error.